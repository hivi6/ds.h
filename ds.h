// ==================================================
// DEFINITIONS
// ==================================================

#ifndef DS_H
#define DS_H

// --------------------------------------------------
// error
// --------------------------------------------------

enum
{
    DS_NO_ERR = 0,
    DS_NULL_ERR,
    DS_MALLOC_ERR,
    DS_RANGE_ERR,
    DS_SIZE_ERR,
    DS_EMPTY_ERR,
    DS_NUM_OF_ERRORS
};

const char *ds_error(int errno);

// --------------------------------------------------
// vector
// --------------------------------------------------

typedef struct
{
    int cap; 	  // capacity of the vector
    int cnt;      // number of items in the vector
    void **items; // array of items
    int *sizes;   // size of each item
} vec_t;

int vec_init(vec_t *vec);
int vec_free(vec_t *vec);
int vec_append(vec_t *vec, void *item, int size);
int vec_pop(vec_t *vec);
int vec_get(vec_t *vec, int index, void *item, int size);
int vec_set(vec_t *vec, int index, void *item, int size);

// --------------------------------------------------
// string builder
// --------------------------------------------------

typedef struct
{
    vec_t chars;
} strb_t;

int strb_init(strb_t *strb);
int strb_free(strb_t *strb);
int strb_append(strb_t *strb, const char *str);
int strb_appendc(strb_t *strb, char ch);
int strb_pop(strb_t *strb);
int strb_get(strb_t *strb, int index, char *ch);
int strb_set(strb_t *strb, int index, char ch);
int strb_build(strb_t *strb, char **str);

#endif // DS_H

// ==================================================
// IMPLEMENTATIONS
// ==================================================

#ifdef DS_IMPLEMENTATION

#include <stdlib.h>
#include <string.h>

// --------------------------------------------------
// error
// --------------------------------------------------

const char *ds_error(int errno)
{
    static const char *msg[] = {
	"No error", // DS_NO_ERR = 0,
	"NULL argument error", // DS_NULL_ERR,
	"Malloc error", // DS_MALLOC_ERR,
	"Index out of range error", // DS_RANGE_ERR,
	"Mismatch size error", // DS_SIZE_ERR,
	"Empty vector error", // DS_EMPTY_ERR,
    };

    if (errno >= DS_NUM_OF_ERRORS)
    {
	return "What is this errno?";
    }
    return msg[errno];
}

// --------------------------------------------------
// vector
// --------------------------------------------------

int vec_init(vec_t *vec)
{
    if (vec == NULL) return DS_NULL_ERR;

    vec->cap = 0;
    vec->cnt = 0;
    vec->items = NULL;
    vec->sizes = NULL;
    return DS_NO_ERR;
}

int vec_free(vec_t *vec)
{
    if (vec == NULL) return DS_NULL_ERR;

    for (int i = 0; i < vec->cnt; i++) free(vec->items[i]);
    free(vec->items);
    free(vec->sizes);

    return vec_init(vec);
}

int vec_append(vec_t *vec, void *item, int size)
{
    if (vec == NULL || item == NULL) return DS_NULL_ERR;

    if (vec->cnt == vec->cap)
    {
	int new_cap = (vec->cap < 8 ? 8 : 2 * vec->cap);
	void **new_items = malloc(new_cap * sizeof(void *));
	int *new_sizes = malloc(new_cap * sizeof(int));
	if (new_items == NULL || new_sizes == NULL) return DS_MALLOC_ERR;

	for (int i = 0; i < vec->cap; i++)
	{
	    new_items[i] = vec->items[i];
	    new_sizes[i] = vec->sizes[i];
	}
	free(vec->items);
	free(vec->sizes);

	vec->cap = new_cap;
	vec->items = new_items;
	vec->sizes = new_sizes;
    }

    char *temp = malloc(sizeof(char) * size);
    if (temp == NULL) return DS_MALLOC_ERR;
    memcpy(temp, item, size);

    vec->items[vec->cnt] = temp;
    vec->sizes[vec->cnt] = size;
    vec->cnt++;
    return DS_NO_ERR;
}

int vec_pop(vec_t *vec)
{
    if (vec == NULL) return DS_NULL_ERR;
    if (vec->cnt == 0) return DS_EMPTY_ERR;

    free(vec->items[vec->cnt - 1]);
    vec->cnt--;
    return DS_NO_ERR;
}

int vec_get(vec_t *vec, int index, void *item, int size)
{
    if (vec == NULL || item == NULL) return DS_NULL_ERR;
    if (index < 0 || index >= vec->cnt) return DS_RANGE_ERR;
    if (vec->sizes[index] != size) return DS_SIZE_ERR;

    memcpy(item, vec->items[index], size);
    return DS_NO_ERR;
}

int vec_set(vec_t *vec, int index, void *item, int size)
{
    if (vec == NULL || item == NULL) return DS_NULL_ERR;
    if (index < 0 || index >= vec->cnt) return DS_RANGE_ERR;

    char *temp = malloc(sizeof(char) * size);
    if (temp == NULL) return DS_MALLOC_ERR;
    memcpy(temp, item, size);

    free(vec->items[index]);
    vec->items[index] = temp;
    vec->sizes[index] = size;
    return DS_NO_ERR;
}

// --------------------------------------------------
// string builder
// --------------------------------------------------

int strb_init(strb_t *strb)
{
    if (strb == NULL) return DS_NULL_ERR;
    return vec_init(&strb->chars);
}

int strb_free(strb_t *strb)
{
    if (strb == NULL) return DS_NULL_ERR;
    return vec_free(&strb->chars);
}

int strb_append(strb_t *strb, const char *str)
{
    if (strb == NULL || str == NULL) return DS_NULL_ERR;
    for (int i = 0; str[i]; i++)
    {
	int errno = vec_append(&strb->chars, (void*)str + i, sizeof(char));
	if (errno) return errno;
    }
    return DS_NO_ERR;
}

int strb_appendc(strb_t *strb, char ch)
{
    if (strb == NULL) return DS_NULL_ERR;
    return vec_append(&strb->chars, &ch, sizeof(ch));
}

int strb_pop(strb_t *strb)
{
    if (strb == NULL) return DS_NULL_ERR;
    return vec_pop(&strb->chars);
}

int strb_get(strb_t *strb, int index, char *ch)
{
    if (strb == NULL || ch == NULL) return DS_NULL_ERR;
    return vec_get(&strb->chars, index, ch, sizeof(*ch));
}

int strb_set(strb_t *strb, int index, char ch)
{
    if (strb == NULL) return DS_NULL_ERR;
    return vec_set(&strb->chars, index, &ch, sizeof(ch));
}

int strb_build(strb_t *strb, char **str)
{
    if (strb == NULL || str == NULL) return DS_NULL_ERR;

    int len = strb->chars.cnt;
    char *temp = malloc(sizeof(char) * len + 1);
    temp[len] = 0;
    for (int i = 0; i < len; i++)
    {	
	int errno = strb_get(strb, i, temp + i);
	if (errno) return errno;
    }

    *str = temp;
    return DS_NO_ERR;
}

#endif // DS_IMPLEMENTATION
