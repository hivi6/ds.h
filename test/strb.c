#include <stdio.h>

#define DS_IMPLEMENTATION
#include "ds.h"

int main()
{
    strb_t strb;
    int errno = strb_init(&strb);
    printf("strb_init() | ERROR: %s\n", ds_error(errno));

    for (int i = 0; i < 26; i++)
    {
	errno = strb_appendc(&strb, 'a' + i);
	printf("strb_appendc(%c) | LEN: %d | ERROR: %s | STR: ", 'a' + i, 
	       strb.chars.cnt, ds_error(errno));
	char *str;
	strb_build(&strb, &str);
	printf("%s\n", str);
	free(str);
    }

    for (int i = 0; strb.chars.cnt; i++)
    {
	for (int j = 0; j < 5; j++)
	{
	    errno = strb_pop(&strb);
	    printf("strb_pop() | LEN: %d | ERROR: %s | STR: ", strb.chars.cnt, 
		   ds_error(errno));
	    char *str;
	    strb_build(&strb, &str);
	    printf("%s\n", str);
	    free(str);
	}
    }

    const char *cstr = "this is a string!";
    strb_append(&strb, cstr);
    char *str;
    errno = strb_build(&strb, &str);
    printf("strb_append(%s) | LEN: %d | ERROR: %s | STR: %s\n", cstr, 
	   strb.chars.cnt, ds_error(errno), str);
    free(str);

    for (int i = 0; strb.chars.cnt; i++)
    {
	for (int j = 0; j < 5; j++)
	{
	    errno = strb_pop(&strb);
	    printf("strb_pop() | LEN: %d | ERROR: %s | STR: ", strb.chars.cnt, 
		   ds_error(errno));
	    char *str;
	    strb_build(&strb, &str);
	    printf("%s\n", str);
	    free(str);
	}
    }

    strb_free(&strb);
    printf("strb_free() | ERROR: %s\n", ds_error(errno));

    return 0;
}
