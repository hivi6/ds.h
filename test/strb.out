strb_init() | ERROR: No error
strb_appendc(a) | LEN: 1 | ERROR: No error | STR: a
strb_appendc(b) | LEN: 2 | ERROR: No error | STR: ab
strb_appendc(c) | LEN: 3 | ERROR: No error | STR: abc
strb_appendc(d) | LEN: 4 | ERROR: No error | STR: abcd
strb_appendc(e) | LEN: 5 | ERROR: No error | STR: abcde
strb_appendc(f) | LEN: 6 | ERROR: No error | STR: abcdef
strb_appendc(g) | LEN: 7 | ERROR: No error | STR: abcdefg
strb_appendc(h) | LEN: 8 | ERROR: No error | STR: abcdefgh
strb_appendc(i) | LEN: 9 | ERROR: No error | STR: abcdefghi
strb_appendc(j) | LEN: 10 | ERROR: No error | STR: abcdefghij
strb_appendc(k) | LEN: 11 | ERROR: No error | STR: abcdefghijk
strb_appendc(l) | LEN: 12 | ERROR: No error | STR: abcdefghijkl
strb_appendc(m) | LEN: 13 | ERROR: No error | STR: abcdefghijklm
strb_appendc(n) | LEN: 14 | ERROR: No error | STR: abcdefghijklmn
strb_appendc(o) | LEN: 15 | ERROR: No error | STR: abcdefghijklmno
strb_appendc(p) | LEN: 16 | ERROR: No error | STR: abcdefghijklmnop
strb_appendc(q) | LEN: 17 | ERROR: No error | STR: abcdefghijklmnopq
strb_appendc(r) | LEN: 18 | ERROR: No error | STR: abcdefghijklmnopqr
strb_appendc(s) | LEN: 19 | ERROR: No error | STR: abcdefghijklmnopqrs
strb_appendc(t) | LEN: 20 | ERROR: No error | STR: abcdefghijklmnopqrst
strb_appendc(u) | LEN: 21 | ERROR: No error | STR: abcdefghijklmnopqrstu
strb_appendc(v) | LEN: 22 | ERROR: No error | STR: abcdefghijklmnopqrstuv
strb_appendc(w) | LEN: 23 | ERROR: No error | STR: abcdefghijklmnopqrstuvw
strb_appendc(x) | LEN: 24 | ERROR: No error | STR: abcdefghijklmnopqrstuvwx
strb_appendc(y) | LEN: 25 | ERROR: No error | STR: abcdefghijklmnopqrstuvwxy
strb_appendc(z) | LEN: 26 | ERROR: No error | STR: abcdefghijklmnopqrstuvwxyz
strb_pop() | LEN: 25 | ERROR: No error | STR: abcdefghijklmnopqrstuvwxy
strb_pop() | LEN: 24 | ERROR: No error | STR: abcdefghijklmnopqrstuvwx
strb_pop() | LEN: 23 | ERROR: No error | STR: abcdefghijklmnopqrstuvw
strb_pop() | LEN: 22 | ERROR: No error | STR: abcdefghijklmnopqrstuv
strb_pop() | LEN: 21 | ERROR: No error | STR: abcdefghijklmnopqrstu
strb_pop() | LEN: 20 | ERROR: No error | STR: abcdefghijklmnopqrst
strb_pop() | LEN: 19 | ERROR: No error | STR: abcdefghijklmnopqrs
strb_pop() | LEN: 18 | ERROR: No error | STR: abcdefghijklmnopqr
strb_pop() | LEN: 17 | ERROR: No error | STR: abcdefghijklmnopq
strb_pop() | LEN: 16 | ERROR: No error | STR: abcdefghijklmnop
strb_pop() | LEN: 15 | ERROR: No error | STR: abcdefghijklmno
strb_pop() | LEN: 14 | ERROR: No error | STR: abcdefghijklmn
strb_pop() | LEN: 13 | ERROR: No error | STR: abcdefghijklm
strb_pop() | LEN: 12 | ERROR: No error | STR: abcdefghijkl
strb_pop() | LEN: 11 | ERROR: No error | STR: abcdefghijk
strb_pop() | LEN: 10 | ERROR: No error | STR: abcdefghij
strb_pop() | LEN: 9 | ERROR: No error | STR: abcdefghi
strb_pop() | LEN: 8 | ERROR: No error | STR: abcdefgh
strb_pop() | LEN: 7 | ERROR: No error | STR: abcdefg
strb_pop() | LEN: 6 | ERROR: No error | STR: abcdef
strb_pop() | LEN: 5 | ERROR: No error | STR: abcde
strb_pop() | LEN: 4 | ERROR: No error | STR: abcd
strb_pop() | LEN: 3 | ERROR: No error | STR: abc
strb_pop() | LEN: 2 | ERROR: No error | STR: ab
strb_pop() | LEN: 1 | ERROR: No error | STR: a
strb_pop() | LEN: 0 | ERROR: No error | STR: 
strb_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
strb_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
strb_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
strb_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
strb_append(this is a string!) | LEN: 17 | ERROR: No error | STR: this is a string!
strb_pop() | LEN: 16 | ERROR: No error | STR: this is a string
strb_pop() | LEN: 15 | ERROR: No error | STR: this is a strin
strb_pop() | LEN: 14 | ERROR: No error | STR: this is a stri
strb_pop() | LEN: 13 | ERROR: No error | STR: this is a str
strb_pop() | LEN: 12 | ERROR: No error | STR: this is a st
strb_pop() | LEN: 11 | ERROR: No error | STR: this is a s
strb_pop() | LEN: 10 | ERROR: No error | STR: this is a 
strb_pop() | LEN: 9 | ERROR: No error | STR: this is a
strb_pop() | LEN: 8 | ERROR: No error | STR: this is 
strb_pop() | LEN: 7 | ERROR: No error | STR: this is
strb_pop() | LEN: 6 | ERROR: No error | STR: this i
strb_pop() | LEN: 5 | ERROR: No error | STR: this 
strb_pop() | LEN: 4 | ERROR: No error | STR: this
strb_pop() | LEN: 3 | ERROR: No error | STR: thi
strb_pop() | LEN: 2 | ERROR: No error | STR: th
strb_pop() | LEN: 1 | ERROR: No error | STR: t
strb_pop() | LEN: 0 | ERROR: No error | STR: 
strb_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
strb_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
strb_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
strb_free() | ERROR: Empty vector error
